exercise/exercise_models.py

```diff
@@ -705,7 +705,7 @@ def grade(self, request, submission, no_penalties=False, url_name="exercise"):
         Loads the exercise feedback page.
         """
         # Get the language from the submission
-        language = submission.lang or get_language()
+        language = submission.lang
```

Comment:
I just noticed that the `submission.lang` can be empty, e.g., for all the currently existing submissions. Rest of the function doesn't work perfectly in that case. I think the `_build_service_url` should be edited to handle empty language values and such. Though, that would require a version of `pick_localized`, which would also return the selected language (it sometimes selects a different than in the language parameter, e.g., when that is not a valid language).

Hmm.. I'm not 100% sure what how to fix this well, but feel free to come up with some ideas.

Reply:
Do you still think it would be bad to save it to meta_data here if None, seems like an ok enough solution here. If it shouldn't be saved is there a reason why changing `_build_servise_url` would be better than `or get_language()` here. Seems like it would make it more complicated for basically the same functionality.

Reply:
Assuming this gets into a async task, then the `get_language()` and `request` are not available, thus the error handling should a) fail the grade or b) find a solution without above methods. In addition, `_build_service_url` already does misbehave with invalid language data, or more exactly, it adds `lang=` query argument which differs from the url it selects as `pick_localized` does pick something (and doesn't tell what) if the requested (invalid) language is not available.

Reply:
I like the simplicity of failing the grade. This will mean that we can't regrade old submissions created before these changes. Is this something we can accept? I'm guessing this problem wouldn't be encountered many times.

Reply:
I would like @Mankro's input for that decision.

Reply:
If `submission.lang` is empty, could we just use the main language of the course? The submission always belongs exactly one course (references are about like this: `submission.exercise.course_module.course_instance`). If the language of the course instance is also empty, then it could default to English. I think it would make sense that the `grade` method selects the language to use in all cases. I don't want to fail grading just because the language is unclear, particularly when course instances usually define one or two languages (course instance language really shouldn't be empty in production even though the database allows empty values there).

I think `_build_service_url` should always receive a valid language parameter.